File: packages/voltaire-ts/src/jsonrpc/eth/getTransactionByHash/eth_getTransactionByHash.go
package eth

import (
	"encoding/json"
	"fmt"

	"github.com/ethereum/execution-apis/types"
)

// Returns the information about a transaction requested by transaction hash.
//
// Example:
// Transaction hash: "0xa52be92809541220ee0aaaede6047d9a6c5d0cd96a517c854d944ee70a0ebb44"
// Result: ...
//
// Implements the eth_getTransactionByHash JSON-RPC method.

// Method is the JSON-RPC method name
const Method = "eth_getTransactionByHash"

// Params represents the parameters for eth_getTransactionByHash
type Params struct {
	// 32 byte hex value
	Transaction hash types.Hash `json:"-"`
}

// MarshalJSON implements json.Marshaler for Params.
// JSON-RPC 2.0 uses positional array parameters.
func (p Params) MarshalJSON() ([]byte, error) {
	return json.Marshal([]interface{}{
		p.Transaction hash,
	})
}

// UnmarshalJSON implements json.Unmarshaler for Params.
func (p *Params) UnmarshalJSON(data []byte) error {
	var arr []json.RawMessage
	if err := json.Unmarshal(data, &arr); err != nil {
		return err
	}

	if len(arr) != 1 {
		return fmt.Errorf("expected 1 parameters, got %d", len(arr))
	}

	if err := json.Unmarshal(arr[0], &p.Transaction hash); err != nil {
		return fmt.Errorf("parameter 0 (Transaction hash): %w", err)
	}

	return nil
}

// Result represents the result for eth_getTransactionByHash
type Result struct {
	Value types.Quantity `json:"-"`
}

// MarshalJSON implements json.Marshaler for Result.
func (r Result) MarshalJSON() ([]byte, error) {
	return json.Marshal(r.Value)
}

// UnmarshalJSON implements json.Unmarshaler for Result.
func (r *Result) UnmarshalJSON(data []byte) error {
	return json.Unmarshal(data, &r.Value)
}

File: packages/voltaire-zig/src/jsonrpc/eth/getTransactionByHash/eth_getTransactionByHash.zig
const std = @import("std");
const types = @import("../../types.zig");

/// Returns the information about a transaction requested by transaction hash.
///
/// Example:
/// Transaction hash: "0xa52be92809541220ee0aaaede6047d9a6c5d0cd96a517c854d944ee70a0ebb44"
/// Result: TransactionResponse or null if not found
///
/// Implements the `eth_getTransactionByHash` JSON-RPC method.
pub const EthGetTransactionByHash = @This();

/// The JSON-RPC method name
pub const method = "eth_getTransactionByHash";

/// Parameters for `eth_getTransactionByHash`
pub const Params = struct {
    /// 32 byte hex value
    transaction_hash: types.Hash,

    pub fn jsonStringify(self: Params, jws: *std.json.WriteStream) !void {
        try jws.beginArray();
        try jws.write(self.transaction_hash);
        try jws.endArray();
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Params {
        if (source != .array) return error.UnexpectedToken;
        if (source.array.items.len != 1) return error.InvalidParamCount;

        return Params{
            .transaction_hash = try std.json.innerParseFromValue(types.Hash, allocator, source.array.items[0], options),
        };
    }
};

/// Result for `eth_getTransactionByHash`
/// Returns TransactionResponse for found transactions, null for not found.
pub const Result = struct {
    value: ?types.TransactionResponse,

    pub fn jsonStringify(self: Result, jws: *std.json.WriteStream) !void {
        if (self.value) |tx| {
            try tx.jsonStringify(jws);
        } else {
            try jws.writeNull();
        }
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Result {
        if (source == .null) {
            return Result{ .value = null };
        }

        // Parse TransactionResponse from JSON
        // For now, we'll parse into a generic structure and create the appropriate variant
        // This is a placeholder - full parsing would need to handle all tx types
        _ = allocator;
        _ = options;
        @panic("TODO: implement TransactionResponse parsing from JSON");
    }
};
